add version to dev builds file names

bcotrim opened the pull request:
## Related issues

- Fixes https://github.com/Automattic/dotcom-forge/issues/7439

## Proposed Changes

- Added version information to development build filenames
- Modified the release manifest generation to include version numbers in URLs
- Updated build naming convention to include semantic version (e.g., v1.2.3-commit_hash)
- Added dry run mode to safely test distribution changes
- Skip external service calls (Slack, Buildkite, S3, Cloudfront) in dry run mode

## Testing Instructions

Note: You need to have all the builds in the `out` folder.

<details>
<summary>expected folder structure</summary>

```bash
out/
├── Studio-darwin-universal/
│   └── Studio.app.zip
├── Studio-darwin-universal.dmg
├── Studio-darwin-x64/
│   └── Studio.app.zip
├── Studio-darwin-x64.dmg
├── Studio-darwin-arm64/
│   └── Studio.app.zip
├── Studio-darwin-arm64.dmg
├── make/
│   └── squirrel.windows/
│       └── x64/
│           ├── studio-setup.exe
└──         └── studio-update.nupkg
```

</details>

1. Test dev build distribution (dry run):

```bash
DRY_RUN=true BUILDKITE_BUILD_NUMBER=test-build bundle exec fastlane distribute_dev_build
```
2. Check the output and confirm files have the expected format

<details>
<summary>Example dry run output</summary>

```bash
[16:02:53]: Running in DRY RUN mode
[16:02:53]: [DRY RUN] Would upload file: Studio-darwin-universal/Studio.app.zip
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-darwin-universal-v1.3.2-0ba91bc.app.zip
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: Studio-darwin-universal.dmg
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-darwin-universal-v1.3.2-0ba91bc.dmg
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: Studio-darwin-x64/Studio.app.zip
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-darwin-x64-v1.3.2-0ba91bc.app.zip
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: Studio-darwin-arm64/Studio.app.zip
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-darwin-arm64-v1.3.2-0ba91bc.app.zip
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: studio-setup.exe
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-win32-v1.3.2-0ba91bc.exe
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: studio-update.nupkg
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/studio-win32-v1.3.2-0ba91bc-full.nupkg
[16:02:53]: if_exists: fail
[16:02:53]: [DRY RUN] Would upload file: releases.json
[16:02:53]: to S3 bucket: a8c-apps-public-artifacts
[16:02:53]: with key: studio/releases.json
[16:02:53]: if_exists: replace
[16:02:53]: [DRY RUN] Would invalidate CloudFront cache:
[16:02:53]: Distribution: EF4A01YASGPY5
[16:02:53]: Paths: /studio/releases.json
[16:02:53]: Commit hash: 0ba91bc-dry-run
```

</details>

3. Run the releases manifest script to check if the dev builds have the right name.

```bash
IS_DEV_BUILD=true node scripts/generate-releases-manifest.mjs
grep -A 20 '"dev"' out/releases.json
```

<details>
<summary>Example releases manifest output</summary>

```json
"dev": {
    "darwin": {
      "universal": {
        "sha": "0ba91bc",
        "url": "https://cdn.a8c-ci.services/studio/studio-darwin-universal-v1.3.2-0ba91bc.app.zip"
      },
      "x64": {
        "sha": "0ba91bc",
        "url": "https://cdn.a8c-ci.services/studio/studio-darwin-x64-v1.3.2-0ba91bc.app.zip"
      },
      "arm64": {
        "sha": "0ba91bc",
        "url": "https://cdn.a8c-ci.services/studio/studio-darwin-arm64-v1.3.2-0ba91bc.app.zip"
      }
    },
    "win32": {
      "sha": "2B249D838747BBC5FF07454961CC4DCF4C05DA4A",
      "url": "https://cdn.a8c-ci.services/studio/studio-win32-v1.3.2-0ba91bc-full.nupkg",
      "size": "199236843"
    }
  },
 ```

</details>

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

bcotrim committed:
add version to dev builds

bcotrim commented:
@Automattic/yolo  👋 

Before finalizing these changes, I'd appreciate some guidance on testing the dev build distribution flow locally. Specifically:

1. How to test the manifest generation with both Windows and macOS builds
2. How to verify that the URLs in `releases.json` match the actual artifact paths in CI
3. Best practices for testing auto-update functionality

I see this involves `prepare-dev-build-version.mjs`, electron-forge builds, and fastlane distribution, but want to ensure I'm testing it thoroughly. Would anyone who's familiar with this process be able to share their testing approach?

Thanks in advance! 🙏

bcotrim committed:
add dry run mode to fastfile

bcotrim commented:
> @Automattic/yolo 👋
> 
> Before finalizing these changes, I'd appreciate some guidance on testing the dev build distribution flow locally. Specifically:
> 
> 1. How to test the manifest generation with both Windows and macOS builds
> 2. How to verify that the URLs in `releases.json` match the actual artifact paths in CI
> 3. Best practices for testing auto-update functionality
> 
> I see this involves `prepare-dev-build-version.mjs`, electron-forge builds, and fastlane distribution, but want to ensure I'm testing it thoroughly. Would anyone who's familiar with this process be able to share their testing approach?
> 
> Thanks in advance! 🙏

Maybe no longer needed as I tried to find a way to test the changes. 
See the PR description above.

fredrikekelund reviewed (approved):
Changes look good and test instructions worked great 👍

For future reference, I had to install `rbenv` (I used Homebrew for that), and tweak the rbenv config (I use oh my zsh, so I added the rbenv plugin and removed the `.zshprofile` file that the installer added). Once that was done, I had to run:
- `rbenv install`
- `rbenv local`
- `gem install bundler`
- `bundle install`

@bcotrim and I discussed how to verify these changes end to end and determined that the right way probably is to install the latest dev build on Windows, merge this PR, and ensure that the auto-updater works after we merged the PR. Waiting for confirmation from @wojtekn about how Squirrel.Windows's auto-updater works…

mokagio commented on fastlane/Fastfile:
```diff
@@ -71,28 +80,61 @@ lane :distribute_release_build do |_options|
+  end
+end
+
+def upload_to_s3_with_dry_run(bucket:, key:, file:, if_exists:, auto_prefix:)
```
Have you considered updating the `upload_to_s3` signature to include a `dry_run:` parameter?

I ask because I don't see any remaining call to the original `upload_to_s3` implementation.

mokagio commented on fastlane/Fastfile:
```diff
@@ -71,28 +80,61 @@ lane :distribute_release_build do |_options|
   )
 end
 
+def clear_cloudfront_cache_with_dry_run(paths:, commit_hash:)
+  if DRY_RUN
+    UI.message("[DRY RUN] Would invalidate CloudFront cache:")
+    UI.message("           Distribution: EF4A01YASGPY5")
```
Same question as `upload_to_s3` vs `upload_to_s3_with_dry_run` here.

As a bonus, it would remove the duplication of the `EF4A01YASGPY5` value across two methods.

mokagio commented on fastlane/Fastfile:
```diff
@@ -189,9 +231,9 @@ def distribute_builds(
       filename = assemble_filename.call(build, 'latest')
       key = "#{bucket_folder}/#{filename}"
 
-      upload_to_s3(
+      upload_to_s3_with_dry_run(
         bucket: bucket_name,
-        key:,
+        key: key,
```
Modern Ruby allows omitting the value in a case like this when the variable name is the same.

By default, the Ruby linter we use raises up the syntax with the value as a style violation. We can turn that off in [the config file](https://github.com/Automattic/studio/blob/e7a67f770ceec8f15105d3cc5aea4ba792ec0fc7/.rubocop.yml), like we've done [in other projects](https://github.com/wordpress-mobile/release-toolkit/blob/ffacf8a24ce7875012ed8d1cacccb1cda2ddf17b/.rubocop.yml#L52-L55).

```yaml
# Rule enforced after migrating to Ruby 3, to prevent the new syntax which allows for hash values and keyword arguments
# omission, such as `myMethod(x:, y:)` and `h = { a:, b: }`.
Style/HashSyntax:
  EnforcedShorthandSyntax: never
```

fredrikekelund commented:
@mokagio, we'd appreciate a review from you on this, too, when you get a chance 🙂

mokagio commented on fastlane/Fastfile:
```diff
@@ -71,28 +80,61 @@ lane :distribute_release_build do |_options|
-      ':studio-app-icon-mac: Release available for': builds.each_value.map { |b| "[#{b[:name]}](#{b[:cdn_url]})" }.join(', ')
-    },
-    default_payloads: []
+  unless DRY_RUN
+    slack(
+      username: 'CI Bot',
+      icon_url: 'https://octodex.github.com/images/jenktocat.jpg',
+      message: ":tada: #{release_tag} is now available! :tada:",
+      channel: '#dotcom-studio',
+      success: true,
+      slack_url: get_required_env('SLACK_WEBHOOK'),
+      payload: {
+        ':studio-app-icon-mac: Release available for': builds.each_value.map { |b| "[#{b[:name]}](#{b[:cdn_url]})" }.join(', ')
+      },
+      default_payloads: []
+    )
+  end
```
What do you think of this syntax which does an early return if in dry run?

```suggestion
  next if DRY_RUN
  
  slack(
    username: 'CI Bot',
    icon_url: 'https://octodex.github.com/images/jenktocat.jpg',
    message: ":tada: #{release_tag} is now available! :tada:",
    channel: '#dotcom-studio',
    success: true,
    slack_url: get_required_env('SLACK_WEBHOOK'),
    payload: {
      ':studio-app-icon-mac: Release available for': builds.each_value.map { |b| "[#{b[:name]}](#{b[:cdn_url]})" }.join(', ')
    },
    default_payloads: []
  )
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -22,6 +29,8 @@ CDN_URL = 'https://cdn.a8c-ci.services'
 # Use this instead of getting values from ENV directly
 # It will throw an error if the requested value is missing
 def get_required_env(key)
+  return 'dry-run' if DRY_RUN && (key == 'BUILDKITE_BUILD_NUMBER' || key == 'BUILDKITE_TAG' || key == 'SLACK_WEBHOOK')
```
What is the use case for this change? I ask because I'm afraid it adds some surprising behavior. I might be overthinking it, but overriding those env var values when in dry run seems a bit strong, what if in dry run the code needs to read the actual value?

mokagio reviewed (commented):
Left a few question on the dry run implementation. The dry run mode is orthogonal to the core of the PR – adding version to dev builds file name – so they don't necessarily need to hold back merging.

> Waiting for confirmation from @wojtekn about how Squirrel.Windows's auto-updater works…

This will be useful to document, as I assume it's not clearly documented in the public docs, so that we can be aware of this in our other apps. Thanks!

wojtekn commented:
> Waiting for confirmation from @wojtekn about how Squirrel.Windows's auto-updater works…

@fredrikekelund I don't have such knowledge. AFAIR it doesn't work for now for dev builds on Windows and this is the issue we aim to fix.

bcotrim committed:
address feedback

bcotrim committed:
address feedback

bcotrim committed:
fix comment

bcotrim commented:
@mokagio Thank you for the review!

I’ve addressed your comments—please let me know if everything looks good or if there’s anything else that needs adjustment.

mokagio reviewed (approved):
Thanks for following up @bcotrim 

Keen to see how this run once merged.
